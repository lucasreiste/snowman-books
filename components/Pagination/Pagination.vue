<template>
  <div class="flex justify-center items-center space-x-4 mt-6">
    <button
      :disabled="currentPage === 1"
      class="px-4 py-2 bg-gray-300 text-gray-600 rounded-md font-medium transition hover:bg-gray-400 disabled:opacity-50 disabled:cursor-not-allowed"
      @click="prevPage"
    >
      <LucideChevronLeft class="w-4 h-4 inline-block mr-1" />
      Anterior
    </button>

    <span class="text-gray-600 font-medium">
      Página {{ currentPage }} de {{ totalPages }}
    </span>

    <button
      :disabled="currentPage === totalPages"
      class="px-4 py-2 bg-gray-300 text-gray-600 rounded-md font-medium transition hover:bg-gray-400 disabled:opacity-50 disabled:cursor-not-allowed"
      @click="nextPage"
    >
      Próxima
      <LucideChevronRight class="w-4 h-4 inline-block ml-1" />
    </button>
  </div>
</template>

<script setup>
const props = defineProps({
  totalItems: {
    type: Number,
    required: true,
  },
  itemsPerPage: {
    type: Number,
    default: 10,
  },
  currentPage: {
    type: Number,
    required: true,
  },
});

const emit = defineEmits(["pageChange"]);

const totalPages = computed(() =>
  Math.ceil(props.totalItems / props.itemsPerPage)
);

const nextPage = () => {
  if (props.currentPage < totalPages.value) {
    emit("pageChange", props.currentPage + 1);
  }
};

const prevPage = () => {
  if (props.currentPage > 1) {
    emit("pageChange", props.currentPage - 1);
  }
};
</script>
