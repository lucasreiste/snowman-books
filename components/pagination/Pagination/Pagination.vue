<template>
  <div class="flex flex-col items-center gap-2">
    <div class="flex items-center gap-2">
      <Button
        variant="outline"
        size="sm"
        :disabled="currentPage === 1"
        @click="prevPage"
      >
        <LucideChevronLeft class="h-4 w-4" />
        <span class="ml-1">Anterior</span>
      </Button>

      <div class="flex items-center gap-1">
        <Button
          v-for="page in totalPages"
          :key="page"
          :variant="page === currentPage ? 'default' : 'outline'"
          size="sm"
          @click="emit('pageChange', page)"
        >
          {{ page }}
        </Button>
      </div>

      <Button
        variant="outline"
        size="sm"
        :disabled="currentPage === totalPages"
        @click="nextPage"
      >
        <span class="mr-1">Pr√≥xima</span>
        <LucideChevronRight class="h-4 w-4" />
      </Button>
    </div>

    <p class="text-sm text-muted-foreground">
      Mostrando {{ startItem }}-{{ endItem }} de {{ totalItems }} itens
    </p>
  </div>
</template>

<script lang="ts" setup>
import { computed } from "vue";
import { LucideChevronLeft, LucideChevronRight } from "lucide-vue-next";
import { Button } from "@/components/ui/button";

interface Props {
  totalItems: number;
  itemsPerPage: number;
  currentPage: number;
}

interface Emits {
  (e: "pageChange", page: number): void;
}

const props = withDefaults(defineProps<Props>(), {
  itemsPerPage: 10,
  currentPage: 1,
});

const emit = defineEmits<Emits>();

const totalPages = computed(() =>
  Math.ceil(props.totalItems / props.itemsPerPage)
);

const startItem = computed(() =>
  props.totalItems === 0 ? 0 : (props.currentPage - 1) * props.itemsPerPage + 1
);

const endItem = computed(() =>
  Math.min(props.currentPage * props.itemsPerPage, props.totalItems)
);

const nextPage = () => {
  if (props.currentPage < totalPages.value) {
    emit("pageChange", props.currentPage + 1);
  }
};

const prevPage = () => {
  if (props.currentPage > 1) {
    emit("pageChange", props.currentPage - 1);
  }
};
</script>
