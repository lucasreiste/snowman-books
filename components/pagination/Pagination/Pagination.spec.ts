import { describe, it, expect } from "vitest";
import { render, screen, fireEvent } from "@testing-library/vue";
import Pagination from "./Pagination.vue";

const ButtonMock = {
  name: "Button",
  template: `<button :data-testid="$attrs['data-testid']" :disabled="disabled" @click="$emit('click')"><slot /></button>`,
  props: ["disabled", "variant", "size"],
  emits: ["click"],
  inheritAttrs: true,
};

describe("Pagination", () => {
  const renderComponent = (props = {}) => {
    return render(Pagination, {
      props: {
        totalItems: 100,
        itemsPerPage: 10,
        currentPage: 1,
        ...props,
      },
      global: {
        stubs: {
          Button: ButtonMock,
          LucideChevronLeft: true,
          LucideChevronRight: true,
        },
      },
    });
  };

  it("should render correctly", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it("should render correctly with default props", () => {
    renderComponent();
    const itemsInfo = screen.getByTestId("items-info");
    expect(itemsInfo).toHaveTextContent("Mostrando 1-10 de 100 itens");
  });

  it("should emit pageChange event when clicking a page", async () => {
    const { emitted } = renderComponent();
    const pageButton = screen.getByTestId("page-button-2");
    await fireEvent.click(pageButton);

    expect(emitted()["pageChange"]).toBeTruthy();
    expect(emitted()["pageChange"][0]).toEqual([2]);
  });

  it("should disable previous button on first page", () => {
    renderComponent();
    const prevButton = screen.getByTestId("prev-button");
    expect(prevButton).toBeDisabled();
  });

  it("should disable next button on last page", () => {
    renderComponent({
      totalItems: 30,
      itemsPerPage: 10,
      currentPage: 3,
    });

    const nextButton = screen.getByTestId("next-button");
    expect(nextButton).toBeDisabled();
  });

  it("should display correct item range for middle page", () => {
    renderComponent({
      totalItems: 100,
      itemsPerPage: 10,
      currentPage: 5,
    });

    const itemsInfo = screen.getByTestId("items-info");
    expect(itemsInfo).toHaveTextContent("Mostrando 41-50 de 100 itens");
  });

  it("should handle zero items correctly", () => {
    renderComponent({
      totalItems: 0,
      itemsPerPage: 10,
      currentPage: 1,
    });

    const itemsInfo = screen.getByTestId("items-info");
    expect(itemsInfo).toHaveTextContent("Mostrando 0-0 de 0 itens");
  });
});
