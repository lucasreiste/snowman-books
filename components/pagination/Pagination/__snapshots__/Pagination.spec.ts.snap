// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > should render correctly 1`] = `
<div>
  <div
    class="flex flex-col items-center gap-2"
  >
    <div
      class="flex items-center gap-2"
    >
      <button
        data-testid="prev-button"
        disabled=""
      >
        
        <svg
          class="lucide lucide-chevron-left-icon h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m15 18-6-6 6-6"
          />
        </svg>
        <span
          class="ml-1"
        >
          Anterior
        </span>
        
      </button>
      <div
        class="flex items-center gap-1"
      >
        
        <button
          data-testid="page-button-1"
        >
          
          1
          
        </button>
        <button
          data-testid="page-button-2"
        >
          
          2
          
        </button>
        <button
          data-testid="page-button-3"
        >
          
          3
          
        </button>
        <button
          data-testid="page-button-4"
        >
          
          4
          
        </button>
        <button
          data-testid="page-button-5"
        >
          
          5
          
        </button>
        <button
          data-testid="page-button-6"
        >
          
          6
          
        </button>
        <button
          data-testid="page-button-7"
        >
          
          7
          
        </button>
        <button
          data-testid="page-button-8"
        >
          
          8
          
        </button>
        <button
          data-testid="page-button-9"
        >
          
          9
          
        </button>
        <button
          data-testid="page-button-10"
        >
          
          10
          
        </button>
        
      </div>
      <button
        data-testid="next-button"
      >
        
        <span
          class="mr-1"
        >
          Pr√≥xima
        </span>
        <svg
          class="lucide lucide-chevron-right-icon h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
        
      </button>
    </div>
    <p
      class="text-sm text-muted-foreground"
      data-testid="items-info"
    >
       Mostrando 1-10 de 100 itens 
    </p>
  </div>
</div>
`;
